### PR TITLE
Inclusion of ANN method

### PR DESCRIPTION
**Pull request by ggurioli**
Last commit with files: Inclusion of ANN method 669c7cd

### CODE REVIEW

File: ezyrb/ann.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+"""
+import torch
+import torch.nn as nn
+import numpy as np
+from .approximation import Approximation
+
+class ANN(Approximation):
+    """
+    Feed-Forward Artifical Neural Network (ANN).
+    : param int trained_epoch: number of already trained iterations.
```
Comment (mtezzele) [inline ezyrb/ann.py:12]: the space are important in the documentation style. Here is :param without spaces. Same in the following lines.

File: tests/test_ann.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    return inp, out
+
+class TestANN(TestCase):
+    def test_constructor_empty(self):
+        ann = ANN()
+
+    def test_fit_mono(self):
+        x, y = get_xy()
+        ann = ANN()
+        ann.fit(x[:, 0].reshape(len(x),1), y[:, 0].reshape(len(y),1))
```
Comment (mtezzele) [inline tests/test_ann.py:28]: just a trick: instead of `reshape(len(x),1)` use `reshape(-1, 1)`. With -1 it will compute the length by itself.

File: ezyrb/ann.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    : param criterion: Loss definition (Mean Squared).
```
Comment (mtezzele) [inline ezyrb/ann.py:13]: You can merge these two lines with `:param torch.nn.modules.loss.MSELoss criterion`

File: ezyrb/ann.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        points = torch.from_numpy(points).float()
+        values = torch.from_numpy(values).float()
+        for epoch in range(niter):
+            y_pred = self.model(points)
+            loss = self.criterion(y_pred, values)
+            self.optimizer.zero_grad()
+            loss.backward()
+            self.optimizer.step()
+        self.trained_epoch += niter
+        return loss.item()
```
Comment (mtezzele) [inline ezyrb/ann.py:70]: is it necessary? Why do you return the loss?

File: ezyrb/ann.py (hunk 1 above)
Comment (ggurioli) [reply ezyrb/ann.py:12]: Thanks, I will fix that.

File: tests/test_ann.py (hunk 2 above)
Comment (ggurioli) [reply tests/test_ann.py:28]: Thanks for the useful hint!

File: ezyrb/ann.py (hunk 3 above)
Comment (ggurioli) [reply ezyrb/ann.py:13]: Thanks for the suggestion!

File: ezyrb/ann.py (hunk 4 above)
Comment (ggurioli) [reply ezyrb/ann.py:70]: It is not necessary, but it could be helpful to have a response on how much effective the training phase is. This can guide the user in the balancing of the net.

Comment (mtezzele) [reply ezyrb/ann.py:70]: it could be a member of the class, for example. Please be consistent with the other fit methods in ezyrb and in general with scikit-learn APIs.

File: ezyrb/ann.py [hunk 5]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
     >>> print(y_pred)
     """
         
-    def __init__(self):
-        self.trained_epoch = 0   
-        self.criterion = torch.nn.MSELoss()
+    def __init__(self, layers, function, stop_training, loss=None):
```
Comment (ggurioli) [inline ezyrb/ann.py:28]: In the current implementation, it is not clear if the "function" input is really needed.

File: ezyrb/ann.py [hunk 6]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
...
+
+        if loss is None: loss = torch.nn.MSELoss
```
Comment (ggurioli) [inline ezyrb/ann.py:30]: loss = torch.nn.MSELoss()

File: ezyrb/ann.py [hunk 7]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
...
+        if optimizer is None: optimizer = torch.optim.Adam
```
Comment (ggurioli) [inline ezyrb/ann.py:31]: "optimizer" has not been introduced yet, I suggest to move this line after line 91.

File: ezyrb/ann.py [hunk 8]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
...
+        return tensor.detach().numpy()
+
+    def _build_model(self):
+        """
+        Build the torch model
+        """
+        self.layers.insert(0, points.shape[1])
+        self.layers.append(values.shape[1]
+        layers = []
+        for i in range(len(layers)-1):
```
Comment (ggurioli) [inline ezyrb/ann.py:64]: It think here there must be "len(self.layers)" in place of "len(layers)".

File: ezyrb/ann.py [hunk 9]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
...
+            layers.append(nn.Linear(self.layers[i], self.layers[i+1]))
+            layers.append(nn.Tanh())
+        layers.append(nn.Linear(self.layers[-2], self.layers[-1]))
```
Comment (ggurioli) [inline ezyrb/ann.py:67]: In my opinion, we can get rid of this line.

File: ezyrb/ann.py [hunk 10]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
...
+        """
+        return torch.from_numpy(array).float()
+
+    def _convert_torch_to_numpy(self, tensor):
+        """
+        Converting data type.
+        """
+        return tensor.detach().numpy()
+
+    def _build_model(self):
```
Comment (ggurioli) [inline ezyrb/ann.py:57]: "points" and "values" should to be considered as input parameters of the method.

File: ezyrb/ann.py [hunk 11]
```diff
@@ -47,28 +85,32 @@ def fit(self, points, values):
...
+        optimizer = torch.optim.Adam(self.model.parameters())
+ 
+        points = self._convert_numpy_to_torch(points)
+        values = self._convert_numpy_to_torch(values)
+
+        n_epoch = 0
+
+        while True:
             y_pred = self.model(points)
             loss = self.criterion(y_pred, values)
```
Comment (ggurioli) [inline ezyrb/ann.py:101]: "self.criterion" is undefined. I would replace it by "self.loss".

File: ezyrb/ann.py [hunk 12]
```diff
@@ -78,7 +120,6 @@ def predict(self, new_point):
         :return: the predicted values via the ANN.
         :rtype: numpy.ndarray
         """
-        new_point = np.array(new_point)
-        new_point = torch.from_numpy(new_point).float()
-        y_new=self.model(new_point)
-        return y_new.detach().numpy()
+        new_point = self._convert_numpy_to_torch(new_point)
```
Comment (ggurioli) [inline ezyrb/ann.py:123]: In order to perform successfully the predict method in case of a list is given as the input new_point (this is, for instance, the case of the example in the tutoral), I suggest to replace this line by "new_point = self._convert_numpy_to_torch(np.array(new_point))".

File: ezyrb/ann.py [hunk 13]
```diff
@@ -25,9 +25,47 @@ class ANN(Approximation):
...
+        Converting data type.
+        """
+        return tensor.detach().numpy()
+
+    def _build_model(self):
+        """
+        Build the torch model
+        """
+        self.layers.insert(0, points.shape[1])
+        self.layers.append(values.shape[1]
```
Comment (ggurioli) [inline ezyrb/ann.py:62]: Missing bracket at the end of this line.

**Comment by ndem0** [comment]
> I think the optimizer should be passed in the init (as you did for the loss for example)

Right now it is not trivial as for the loss: the optimizer has to be constructed passing the model parameters, which means:
- we can pass the class instead the instance, but the optimizer arguments are then the default ones (or you have to create a new class with the wanted parameters)
- we can pass a dictionary with the optimizer class and all the arguments.

I prefer the second option, we plan to insert something like that in future
